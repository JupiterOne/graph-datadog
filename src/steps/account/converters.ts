import { Organization } from '@datadog/datadog-api-client/dist/packages/datadog-api-client-v1';
import {
  createIntegrationEntity,
  parseTimePropertyValue,
} from '@jupiterone/integration-sdk-core';
import { Entities } from '../constants';

/*
  Modifies the key to not contain spaces/upper characters

  "_key": "datadog_account:Local Integration"
  -> to
  "_key": "datadog_account:local-integration"
*/
function getAccountKey(id: string): string {
  return `datadog_account:${id.toLowerCase().split(/\s+/).join('-')}`;
}

export function createAccountEntity(instanceName: string) {
  return createIntegrationEntity({
    entityData: {
      source: {
        name: instanceName,
      },
      assign: {
        _class: Entities.ACCOUNT._class,
        _type: Entities.ACCOUNT._type,
        _key: getAccountKey(instanceName),
        name: instanceName,
        displayName: instanceName,
      },
    },
  });
}

function getOrganizationAccountKey(id: string): string {
  return `datadog_account:${id}`;
}

export function createOrganizationAccountEntity(account: Organization) {
  return createIntegrationEntity({
    entityData: {
      source: account,
      assign: {
        _class: Entities.ACCOUNT._class,
        _type: Entities.ACCOUNT._type,
        _key: getOrganizationAccountKey(account.publicId!),
        name: account.name,
        displayName: account.name,
        description: account.description ?? '',
        createdOn: parseTimePropertyValue(account.created),
        billingType: account.billing?.type,
        subscriptionType: account.subscription?.type,
        'settings.privateWidgetShare': account.settings?.privateWidgetShare,
        'settings.samlEnabled': account.settings?.saml?.enabled,
        'settings.samlAutocreateAccessRole': account.settings
          ?.samlAutocreateAccessRole as 'st' | 'adm' | 'ro' | 'ERROR',
        'settings.samlAutocreateUsersDomainsEnabled':
          account.settings?.samlAutocreateUsersDomains?.enabled,
        'settings.samlCanBeEnabled': account.settings?.samlCanBeEnabled,
        'settings.samlIdpEndpoint': account.settings?.samlIdpEndpoint,
        'settings.samlIdpInitiatedLoginEnabled':
          account.settings?.samlIdpInitiatedLogin?.enabled,
        'settings.samlIdpMetadataUploaded':
          account.settings?.samlIdpMetadataUploaded,
        'settings.samlLoginUrl': account.settings?.samlLoginUrl,
        'settings.samlStrictModeEnabled':
          account.settings?.samlStrictMode?.enabled,
      },
    },
  });
}
