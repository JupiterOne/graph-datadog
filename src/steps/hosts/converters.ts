import {
  createIntegrationEntity,
  Entity,
  parseTimePropertyValue,
} from '@jupiterone/integration-sdk-core';

import { Entities } from '../constants';

export function getHostKey(id: number): string {
  return `datadog_host:${id}`;
}

// https://docs.datadoghq.com/api/latest/hosts/

/**
 * Creates host entity
 * Does NOT use the Host.d.ts model because of a bug in the package.
 * node-fetch is used to get this data.
 * @param host
 */
export function createHostEntity(host): Entity {
  const osVersion = determineOsVersion({
    macV: host.meta?.macV,
    winV: host.meta?.winV,
    nixV: host.meta?.nixV,
  });

  return createIntegrationEntity({
    entityData: {
      source: host,
      assign: {
        _type: Entities.HOST._type,
        _class: Entities.HOST._class,
        _key: getHostKey(host.id!),
        id: `${host.id}`,
        aliases: host.aliases,
        apps: host.apps,
        awsName: host.aws_name,
        hostname: host.host_name,
        isMuted: host.is_muted,
        reportedOn: parseTimePropertyValue(host.last_reported_time),
        name: host.name,
        sources: host.sources,
        state: host.up ? 'running' : 'stopped',
        agentVersion: host.meta?.agent_version,
        hostCpuCores: host.meta?.cpuCores,
        machine: host.meta?.machine,
        platform: host.meta?.platform,
        processor: host.meta?.processor,
        pythonVersion: host.meta?.pythonV,
        socketFqdn: host.meta?.['socket-fqdn'],
        socketHostname: host.meta?.['socket-hostname'],
        osName: host.meta?.platform,
        osVersion,
      },
    },
  });
}

/**
 * Determines the OS Version details
 * Example data:
 *  [ '13.1', [ '', '', '' ], 'amd64' ]
 *  [ 'ubuntu', '22.04', '' ]
 * @param macV
 * @param winV
 * @param nixV
 */
export function determineOsVersion({ macV, winV, nixV }) {
  if (Array.isArray(macV) && macV.filter((prop) => prop).length > 0) {
    return reduceOsVersion(macV);
  } else if (Array.isArray(winV) && winV.filter((prop) => prop).length > 0) {
    return reduceOsVersion(winV);
  } else if (Array.isArray(nixV) && nixV.filter((prop) => prop).length > 0) {
    return reduceOsVersion(nixV);
  }
}

/**
 * Produces a single string that includes all provided data.
 * @param version
 */
function reduceOsVersion(version) {
  return version
    .filter((prop) => prop)
    .map((prop) => {
      if (Array.isArray(prop)) {
        return prop.filter((prop) => prop).join(' ');
      } else {
        return prop;
      }
    })
    .filter((prop) => prop)
    .join(' ');
}
