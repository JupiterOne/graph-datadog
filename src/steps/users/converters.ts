import { User } from '@datadog/datadog-api-client/dist/packages/datadog-api-client-v2/models/User';
import {
  createIntegrationEntity,
  Entity,
  parseTimePropertyValue,
} from '@jupiterone/integration-sdk-core';

import { Entities } from '../constants';

export function getUserKey(email: string, id: string): string {
  return `datadog_user:${email}:${id}`;
}

export function createUserEntity(user: User): Entity | null {
  return createIntegrationEntity({
    entityData: {
      source: user,
      assign: {
        _type: Entities.USER._type,
        _class: Entities.USER._class,
        _key: getUserKey(user.attributes?.email as string, user.id as string),
        createdOn: parseTimePropertyValue(user.attributes?.createdAt),
        updatedOn: parseTimePropertyValue(user.attributes?.modifiedAt),
        username: user.attributes?.handle,
        email: user.attributes?.email,
        id: user.id,
        name: user.attributes?.name || user.attributes?.email,
        verified: user.attributes?.verified,
        disabled: user.attributes?.disabled,
        active: !user.attributes?.disabled,
        icon: user.attributes?.icon,
        displayName: user.attributes?.name || user.attributes?.email,
      },
    },
  });
}
