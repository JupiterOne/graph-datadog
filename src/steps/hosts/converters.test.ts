import { Host } from '@datadog/datadog-api-client/dist/packages/datadog-api-client-v1/models/Host';
import { getMockHost } from '../../../test/mocks';
import { createHostEntity, determineOsVersion } from './converters';

describe('#createHostEntity', () => {
  test('should convert to entity', () => {
    expect(createHostEntity(getMockHost() as Host)).toMatchSnapshot();
  });

  test('#determineOsVersion', () => {
    expect(
      determineOsVersion({
        macV: ['13.1', ['', '', ''], 'amd64'],
        winV: undefined,
        nixV: null,
      }),
    ).toBe('13.1 amd64');

    expect(
      determineOsVersion({
        macV: undefined,
        winV: ['Windows'],
        nixV: [[], [''], [null, undefined, '']],
      }),
    ).toBe('Windows');

    expect(
      determineOsVersion({
        macV: undefined,
        winV: ['', null, undefined],
        nixV: ['ubuntu', '22.04', ''],
      }),
    ).toBe('ubuntu 22.04');
  });
});
