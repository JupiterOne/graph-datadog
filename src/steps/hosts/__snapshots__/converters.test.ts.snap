// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#createHostEntity should convert to entity 1`] = `
Object {
  "_class": Array [
    "Host",
  ],
  "_key": "datadog_host:9231106664",
  "_rawData": Array [
    Object {
      "name": "default",
      "rawData": Object {
        "aliases": Array [
          "mindfield",
        ],
        "apps": Array [
          "agent",
          "container",
          "docker",
          "ntp",
        ],
        "hostName": "mindfield",
        "id": 9231106664,
        "isMuted": false,
        "lastReportedTime": 1670945215,
        "meta": Object {
          "agentVersion": "7.40.1",
          "cpuCores": 1,
          "fbsdV": Array [
            "",
            "",
            "",
          ],
          "gohai": "{\\"cpu\\":{\\"cache_size\\":\\"512 KB\\",\\"cpu_cores\\":\\"8\\",\\"cpu_logical_processors\\":\\"16\\",\\"family\\":\\"23\\",\\"mhz\\":\\"2200.000\\",\\"model\\":\\"8\\",\\"model_name\\":\\"AMD Ryzen 7 2700X Eight-Core Processor\\",\\"stepping\\":\\"2\\",\\"vendor_id\\":\\"AuthenticAMD\\"},\\"filesystem\\":[{\\"kb_size\\":\\"950914724\\",\\"mounted_on\\":\\"/\\",\\"name\\":\\"overlay\\"},{\\"kb_size\\":\\"65536\\",\\"mounted_on\\":\\"/dev\\",\\"name\\":\\"tmpfs\\"},{\\"kb_size\\":\\"65536\\",\\"mounted_on\\":\\"/dev/shm\\",\\"name\\":\\"shm\\"},{\\"kb_size\\":\\"950914724\\",\\"mounted_on\\":\\"/etc/hosts\\",\\"name\\":\\"/dev/nvme1n1p2\\"},{\\"kb_size\\":\\"1630628\\",\\"mounted_on\\":\\"/var/run/docker.sock\\",\\"name\\":\\"tmpfs\\"},{\\"kb_size\\":\\"8153124\\",\\"mounted_on\\":\\"/proc/asound\\",\\"name\\":\\"tmpfs\\"},{\\"kb_size\\":\\"8153124\\",\\"mounted_on\\":\\"/proc/acpi\\",\\"name\\":\\"tmpfs\\"},{\\"kb_size\\":\\"8153124\\",\\"mounted_on\\":\\"/proc/scsi\\",\\"name\\":\\"tmpfs\\"},{\\"kb_size\\":\\"8153124\\",\\"mounted_on\\":\\"/sys/firmware\\",\\"name\\":\\"tmpfs\\"}],\\"memory\\":{\\"swap_total\\":\\"9229308kB\\",\\"total\\":\\"16306252kB\\"},\\"network\\":null,\\"platform\\":{\\"GOOARCH\\":\\"amd64\\",\\"GOOS\\":\\"linux\\",\\"goV\\":\\"1.18.8\\",\\"hardware_platform\\":\\"x86_64\\",\\"hostname\\":\\"f992d058a188\\",\\"kernel_name\\":\\"Linux\\",\\"kernel_release\\":\\"5.15.0-56-generic\\",\\"kernel_version\\":\\"#62-Ubuntu SMP Tue Nov 22 19:54:14 UTC 2022\\",\\"machine\\":\\"x86_64\\",\\"os\\":\\"GNU/Linux\\",\\"processor\\":\\"x86_64\\",\\"pythonV\\":\\"3.8.14\\"}}",
          "installMethod": Object {
            "installerVersion": "docker",
            "tool": "docker",
            "toolVersion": "docker",
          },
          "macV": Array [
            "",
            "",
            "",
          ],
          "machine": "amd64",
          "nixV": Array [
            "ubuntu",
            "22.04",
            "",
          ],
          "platform": "linux",
          "processor": "AMD Ryzen 7 2700X Eight-Core Processor",
          "pythonV": "3.8.14",
          "socketFqdn": "f992d058a188",
          "socketHostname": "f992d058a188",
          "winV": Array [
            "",
            "",
            "",
          ],
        },
        "metrics": Object {
          "cpu": 6.804911,
          "iowait": 0.00046491018,
          "load": 0.093782894,
        },
        "muteTimeout": undefined,
        "name": "mindfield",
        "sources": Array [
          "agent",
        ],
        "tagsBySource": Object {
          "Datadog": Array [
            "host:mindfield",
          ],
        },
        "up": true,
      },
    },
  ],
  "_type": "datadog_host",
  "agentVersion": "7.40.1",
  "aliases": Array [
    "mindfield",
  ],
  "apps": Array [
    "agent",
    "container",
    "docker",
    "ntp",
  ],
  "awsName": undefined,
  "createdOn": undefined,
  "displayName": "mindfield",
  "hostCpuCores": 1,
  "hostname": "mindfield",
  "id": "9231106664",
  "isMuted": false,
  "machine": "amd64",
  "name": "mindfield",
  "osName": "linux",
  "osVersion": "ubuntu 22.04",
  "platform": "linux",
  "processor": "AMD Ryzen 7 2700X Eight-Core Processor",
  "pythonVersion": "3.8.14",
  "reportedOn": 1670945215,
  "socketFqdn": "f992d058a188",
  "socketHostname": "f992d058a188",
  "sources": Array [
    "agent",
  ],
  "state": "running",
}
`;
